platform/logger.cpp

```diff
@@ -0,0 +1,273 @@
+            core_util_critical_section_exit();
+            if (false == in_isr) {
+                mbed_log_unlock();
+            }
+            return;
+        }
+        int32_t bytes_written = 0;
+        while (count--) {
+            log_buffer_->push(buf[bytes_written++]);
+        }
+        core_util_critical_section_exit();
+        if (false == in_isr) {
+            mbed_log_unlock();
+        }
+    }
+}
+
+#else // String based implementation
+
+static void log_format_str_data(const char *format, va_list args, bool in_isr)
+{
+    LOG_DATA_TYPE_ one_line[LOG_SINGLE_STR_SIZE_ + LOG_SINGLE_HELPER_STR_SIZE_];
+    int32_t size = LOG_SINGLE_STR_SIZE_ + LOG_SINGLE_HELPER_STR_SIZE_;
+
+    volatile uint64_t time = ticker_read_us(log_ticker_);
+    int32_t count = 0;
+    if (time_enable_) {
+        count = snprintf(one_line, size, "[%-8lld]", time);
+        if (count < 0 || count > size) {
+            return;
+        }
+    }
+    count += vsnprintf(one_line + count, (size - count), format, args);
+    if (count > size) {
+        one_line[size-3] = '*';
+        one_line[size-2] = '\n';
+        one_line[size-1] = '\0';
+        count = size;
+    }
+    if (NULL != extern_buf_) {
+        log_update_buf(one_line, count);
+    } else {
+        if (true == in_isr) {
+#ifndef MBED_DISABLE_ISR_LOGGING
+            count += 1;
+            // Check buf size, if we have space for full line then only push
+            if ( (LOG_BUF_SIZE_ - log_buffer_->size()) < count) {
+                return;
+            }
+            int32_t bytes_written = 0;
+            while (count--) {
+                log_buffer_->push(one_line[bytes_written++]);
+            }
+#endif
+        } else {
+            fputs(one_line, stderr);
+        }
+    }
+}
+
+#ifndef MBED_DISABLE_ISR_LOGGING
+// Note: ISR data is lossy, and with too many interrupts it might be
+// over-written and last valid ISR data will be printed.
+extern "C" void print_buf_data(void)
+{
+    int32_t count = 0;
+    LOG_DATA_TYPE_ one_line[LOG_SINGLE_STR_SIZE_];
+    while (1) {
+        while (!log_buffer_->empty()) {
+            mbed_log_lock();
+            log_buffer_->pop(one_line[count]);
+            if ('\0' == one_line[count]) {
+                fputs(one_line, stderr);
+                count = 0;
+            } else {
+                count++;
+                if (count >= LOG_SINGLE_STR_SIZE_) {
+                    count = 0;
+                }
+            }
+            mbed_log_unlock();
+        }
+        Thread::wait(1);
+    }
+}
+#endif
+
+extern "C" void log_str_data(const char *format, ...)
+{
+    va_list args;
+    va_start(args, format);
+    log_str_vdata(format, args);
+    va_end(args);
+}
+
+extern "C" void log_str_vdata(const char *format, va_list args)
+{
+    MBED_STATIC_ASSERT(LOG_SINGLE_STR_SIZE_ >= 32, "String size too small, min 32 bytes");
+
+    if (core_util_is_isr_active() || !core_util_are_interrupts_enabled()) {
```

Comment:
Would it be possible to not stop the world for the format operation and only stop it when the string created locally are copied into the buffer ?